### PR TITLE
Support for state migration to checks state storage and new dump API

### PR DESCRIPTION
**Pull request by KimKyungup**
## Proposed changes

This PR adds the feature to check state/storage trie and adds new dump API.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [ ] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [ ] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: blockchain/state/iterator.go [hunk 1]
```diff
@@ -38,10 +39,13 @@ type NodeIterator struct {
 
 	accountHash common.Hash // Hash of the node containing the account
 	codeHash    common.Hash // Hash of the contract source code
+	addrHash    common.Hash // Hash of address of the account
```
Comment (winnie-byun) [inline blockchain/state/iterator.go:42]: What is `addrHash` for? It is not used.

File: blockchain/state/iterator.go [hunk 2]
```diff
@@ -148,14 +153,98 @@ func (it *NodeIterator) retrieve() bool {
...
+			it.Type = "state_leaf"
+		}
+
+		it.Hash, it.Parent, it.Path = it.stateIt.Hash(), it.stateIt.Parent(), it.stateIt.Path()
 	}
 	return true
 }
+
+// CheckStateConsistency checks the consistency of all state/storage trie of given two state database.
+func CheckStateConsistency(oldDB database.DBManager, newDB database.DBManager, root common.Hash) error {
```
Comment (winnie-byun) [inline blockchain/state/iterator.go:181]: What do you think of using`reflect.DeepEqual` for checking?

File: blockchain/state/iterator.go [hunk 3]
```diff
@@ -148,14 +153,98 @@ func (it *NodeIterator) retrieve() bool {
 	// Otherwise retrieve the current entry
 	switch {
 	case it.dataIt != nil:
-		it.Hash, it.Parent = it.dataIt.Hash(), it.dataIt.Parent()
+		it.Type = "storage"
```
Comment (winnie-byun) [inline blockchain/state/iterator.go:156]: How about using `it.Type`'s own enum type?

File: blockchain/state/iterator.go (hunk 1 above)
Comment (KimKyungup) [reply blockchain/state/iterator.go:42]: It can be removed. I will. :)

File: node/cn/api.go [hunk 4]
```diff
@@ -240,6 +240,49 @@ func (api *PublicDebugAPI) DumpBlock(blockNr rpc.BlockNumber) (state.Dump, error
...
+}
+
+type DumpStateTrieResult struct {
+	Root  string `json:"root"`
+	Tries []Trie `json:"tries"`
+}
+
+// DumpStateTrie retrieves all state/storage tries of the given state root.
+func (api *PublicDebugAPI) DumpStateTrie(blockNr uint64) (DumpStateTrieResult, error) {
+	//func (api *PublicDebugAPI) DumpStateTrie(blockNr rpc.BlockNumber) error {
```
Comment (winnie-byun) [inline node/cn/api.go:257]: Any reason why there is a comment? Also, `rpc.BlockNumber` might be better idea then `uint64`.

File: blockchain/state/iterator.go (hunk 3 above)
Comment (KimKyungup) [reply blockchain/state/iterator.go:155]: This `type` will be printed out in API.
So if we can use enum type, we need additional code to convert string.
I think this is enough. :)

File: node/cn/api.go (hunk 4 above)
Comment (KimKyungup) [reply node/cn/api.go]: I removed.
`rpc.BlockNumber` doesn't support decimal, supports only hex.
So I changed it.

File: blockchain/state/iterator.go (hunk 2 above)
Comment (KimKyungup) [reply blockchain/state/iterator.go:180]: `reflect.DeepEqual` is better? :)

Comment (winnie-byun) [reply blockchain/state/iterator.go:180]: If you are checking all of the variables, it is could be better. It makes sure there is no error in our code.

Comment (winnie-byun) [reply blockchain/state/iterator.go:180]: `state.NodeIterator` has pointers which leads to unequality. I will take my suggestion back. Thank you for considering.

File: storage/statedb/encoding.go [hunk 5]
```diff
@@ -119,3 +123,13 @@ func prefixLen(a, b []byte) int {
 func hasTerm(s []byte) bool {
 	return len(s) > 0 && s[len(s)-1] == 16
 }
+
+// HexKeyPathToHashString turns hex nibbles of trie path key into hash string of the trie path key.
+func HexKeyPathToHashString(path []byte) string {
```
Comment (winnie-byun) [inline storage/statedb/encoding.go:128]: Why do you turn `Path` to `Hash`? Isn't it better if we can how the nodes are connected?

Comment (KimKyungup) [reply storage/statedb/encoding.go:128]: Path is bytes. Hash is ascii. 
So hash is more short and understandable. :)
This is an example. 
`10101010 -> 0xAAAA`

**Comment by KimKyungup** [comment]
PTAL. :)

File: storage/statedb/encoding.go (hunk 5 above)
Comment (winnie-byun) [reply storage/statedb/encoding.go:128]: To me, the word `Hash` sounds like you hashed the path.
Could you change the function name to `HexPathToString` or `HexKeyPathToEncodedString`?
```suggestion
// HexPathToString turns hex nibbles of trie path key into string of the trie path key.
func HexPathToString(path []byte) string {
```

**Pull request merged**